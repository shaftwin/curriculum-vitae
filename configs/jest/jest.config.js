const path = require('path');

module.exports = {
  rootDir: '../../',
  testEnvironment: 'jsdom',
  testMatch: ['**/*.test.js'],
  setupFilesAfterEnv: ['./node_modules/@testing-library/jest-dom'],
  moduleDirectories: [path.resolve(__dirname, '../../'), 'node_modules'],
  setupFiles: [
    './configs/jest/mock.js',
    './configs/cucumber/jest-cucumber-config.js',
  ],
  snapshotSerializers: ['@emotion/jest/serializer'],
};
