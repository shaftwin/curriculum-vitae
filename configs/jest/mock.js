jest.mock('react-i18next', () => ({
  withTranslation: () => (Component) => {
    // eslint-disable-next-line no-param-reassign
    Component.defaultProps = { ...Component.defaultProps, t: () => '' };
    return Component;
  },
  // this mock makes sure any components using the
  // translate hook can use it without a warning being shown
  useTranslation: () => {
    return {
      t: (str) => str,
      i18n: {
        changeLanguage: () => new Promise(() => {}),
      },
    };
  },
}));
