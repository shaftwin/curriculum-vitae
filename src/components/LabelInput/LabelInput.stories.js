import React from 'react';
import LabelInput from './index';

export default {
  title: 'Components/LabelInput',
  component: LabelInput,
};

// eslint-disable-next-line react/jsx-props-no-spreading
const Template = (args) => <LabelInput {...args} />;

export const Default = Template.bind({});
Default.args = {
  label: 'LabelInput',
  inputValue: 'test value',
};

export const LargeInput = Template.bind({});
LargeInput.args = {
  label: 'LargeInput',
  inputValue: 'test value',
  size: 'large',
};

export const InputWithErrorMessage = Template.bind({});
InputWithErrorMessage.args = {
  label: 'LargeInput',
  inputValue: 'test value',
  errorMessage: 'this is an error',
  size: 'large',
  inputState: 'error',
};

export const InputWithoutErrorMessage = Template.bind({});
InputWithoutErrorMessage.args = {
  label: 'LargeInput',
  inputValue: 'test value',
  size: 'large',
  inputState: 'error',
};

export const TextArea = Template.bind({});
TextArea.args = {
  label: 'TextArea',
  inputValue: 'test value',
  textArea: true,
};
