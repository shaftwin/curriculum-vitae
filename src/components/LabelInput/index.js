import React, { useCallback, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import Error from '../../../assets/svg/error';
import {
  Container,
  CustomText,
  InputWrapper,
  CustomInput,
  CustomTextArea,
  ErrorMessage,
  IconContainer,
} from './styled';

// Built-in props are only spread inside the input
const LabelInput = ({
  label,
  inputValue,
  errorMessage,
  inputState,
  type,
  size,
  textArea,
  register,
  className,
  ...props
}) => {
  const [value, setValue] = useState(inputValue);
  const [state, setState] = useState(inputState);
  const handleChange = useCallback((event) => setValue(event.target.value), []);
  const InputComponent = textArea ? CustomTextArea : CustomInput;

  const updateState = useCallback(() => {
    switch (state) {
      case 'default':
        if (value) {
          setState('filled');
        }
        break;
      case 'filled':
        if (!value) {
          setState('default');
        }
        break;
      default:
        break;
    }
  }, [state, value]);

  useEffect(() => {
    updateState();
  }, [updateState]);

  return (
    <Container data-testid="LabelInput" size={size} className={className}>
      {!!label && (
        <CustomText data-testid="label" state={state}>
          {label}
        </CustomText>
      )}
      <InputWrapper
        data-testid="inputWrapper"
        textArea={textArea}
        disabled={state === 'disabled'}
        size={size}
        state={state}
      >
        <InputComponent
          data-testid="input"
          value={value}
          onChange={handleChange}
          type={type}
          ref={register}
          disabled={state === 'disabled'}
          // eslint-disable-next-line react/jsx-props-no-spreading
          {...props}
        />

        {state === 'error' && (
          <IconContainer
            data-testid="errorIcon"
            size={size}
            textArea={textArea}
          >
            <Error />
          </IconContainer>
        )}
      </InputWrapper>
      {state === 'error' && !!errorMessage && (
        <ErrorMessage data-testid="errorMessage" state={state}>
          {errorMessage}
        </ErrorMessage>
      )}
    </Container>
  );
};

LabelInput.propTypes = {
  /**
   * Optional label
   */
  label: PropTypes.string,
  /**
   * Optional inputValue
   */
  inputValue: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /**
   * Optional errorMessage
   */
  errorMessage: PropTypes.string,
  /**
   * Optional state
   * hover and focused will be managed directly in the style file
   */
  inputState: PropTypes.oneOf(['default', 'disabled', 'error']),
  /**
   * Optional type
   */
  type: PropTypes.oneOf(['text', 'number']),
  /**
   * Optional size
   */
  size: PropTypes.oneOf(['medium', 'large']),
  /**
   * Optional textArea
   */
  textArea: PropTypes.bool,
  /**
   * Optional register in order to use react hook form
   */
  register: PropTypes.shape({}),
  /**
   * Optional custom style
   */
  className: PropTypes.string,
};

LabelInput.defaultProps = {
  label: null,
  inputValue: '',
  errorMessage: null,
  inputState: 'default',
  type: 'text',
  size: 'medium',
  textArea: false,
};

export default LabelInput;
