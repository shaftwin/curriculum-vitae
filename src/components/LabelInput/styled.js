import styled from '@emotion/styled';
import { InputStyle } from '../../../utils/sharedStyle/InputStyle';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  font-size: ${({ theme }) => theme.global.fontSize[13]};
  width: ${({ theme, size }) => theme.labelInput.width[size]};
`;

export const InputWrapper = styled(InputStyle)`
  display: flex;
  align-items: ${({ textArea }) => (textArea ? 'baseline' : 'center')};
  padding: ${({ textArea }) => (textArea ? '0.5rem 0.75rem' : '0rem 0.75rem')};
`;

export const CustomInput = styled.input`
  outline: none;
  border: 0;
  background-color: inherit;
  color: inherit;
  width: 100%;
  font-size: ${({ theme }) => theme.global.fontSize[13]};
  min-height: 2rem;
  cursor: inherit;
`;

export const CustomTextArea = styled.textarea`
  outline: none;
  border: 0;
  background-color: inherit;
  color: inherit;
  font-size: ${({ theme }) => theme.global.fontSize[13]};
  cursor: inherit;
  min-height: 4.5rem;
  flex: 1;
`;

export const CustomText = styled.label`
  margin-bottom: 0.25rem;
  font-weight: ${({ theme }) => theme.global.fontWeight.semiBold};
  color: ${({ theme, state }) =>
    state === 'error'
      ? theme.global.colors.status.error
      : theme.global.colors.darkBlue};
`;

export const ErrorMessage = styled(CustomText)`
  margin: 0.25rem 0rem 0rem 0rem;
  font-size: ${({ theme }) => theme.global.fontSize[10]};
`;

export const IconContainer = styled.div`
  position: ${({ textArea }) => (textArea ? 'absolute' : 'relative')};
  left: ${({ theme, size, textArea }) =>
    textArea ? theme.labelInput.iconLeft[size] : 'unset'};
`;
