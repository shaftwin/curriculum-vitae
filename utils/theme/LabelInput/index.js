import { global } from '../global';

export const labelInput = {
  border: {
    default: global.colors.grey,
    filled: global.colors.mediumGrey,
    disabled: global.colors.grey,
    hover: global.colors.grey,
    error: global.colors.status.error,
    focused: global.colors.green,
  },
  background: {
    disabled: global.colors.grey,
  },
  color: {
    default: global.colors.mediumGrey,
    hover: global.colors.mediumGrey,
    focused: global.colors.brownGrey,
    filled: global.colors.brownGrey,
    error: global.colors.brownGrey,
    disabled: global.colors.disabledGrey,
  },
  width: {
    medium: '16.625rem',
    large: '21rem',
  },
  iconLeft: {
    medium: '15.5rem',
    large: '19.75rem',
  },
};
