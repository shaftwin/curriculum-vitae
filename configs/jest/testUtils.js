// import { MockedProvider } from '@apollo/client/testing';
import { ThemeProvider } from '@emotion/react';
import { render } from '@testing-library/react';
import React from 'react';
import theme from '../../utils/theme';

const AllTheProviders = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
