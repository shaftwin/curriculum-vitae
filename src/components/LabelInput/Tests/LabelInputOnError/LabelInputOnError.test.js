import * as React from 'react';
import { screen } from '@testing-library/react';
import { defineFeature, loadFeature } from 'jest-cucumber';
import { render } from '../../../../../configs/jest/testUtils';
import LabelInput from '../..';

const feature = loadFeature(
  'src/components/LabelInput/Tests/LabelInputOnError/LabelInputOnError.feature',
);

defineFeature(feature, (test) => {
  let container;

  // TODO: Externalize this method
  const givenLabelInputIsRendered = (given, comp) => {
    given('LabelInput component is rendered', () => {
      container = render(comp).container;
    });
  };

  test('Default LabelInput on error without message', ({ given, and }) => {
    givenLabelInputIsRendered(
      given,
      <LabelInput
        label="testLabel"
        inputValue="testValue"
        inputState="error"
      />,
    );

    and('Required elements are mounted', () => {
      expect(screen.getByTestId('LabelInput')).toBeInTheDocument();
      expect(screen.getByTestId('input')).toBeInTheDocument();
      expect(screen.getByText(/testLabel/i)).toBeInTheDocument();
      expect(screen.getByTestId('inputWrapper')).toBeInTheDocument();
      expect(screen.getByTestId('errorIcon')).toBeInTheDocument();
    });

    and('Undesirable elements are not mounted', () => {
      expect(screen.queryByTestId('errorMessage')).not.toBeInTheDocument();
    });

    and('LabelInput should have the error style', () => {
      expect(container).toMatchSnapshot();
    });
  });

  test('LabelInput throw error with message', ({ given, and }) => {
    givenLabelInputIsRendered(
      given,
      <LabelInput
        label="testLabel"
        inputValue="testValue"
        inputState="error"
        errorMessage="errorMessage"
      />,
    );

    and('Required elements are mounted', () => {
      expect(screen.getByTestId('LabelInput')).toBeInTheDocument();
      expect(screen.getByTestId('input')).toBeInTheDocument();
      expect(screen.getByText(/testLabel/i)).toBeInTheDocument();
      expect(screen.getByTestId('inputWrapper')).toBeInTheDocument();
      expect(screen.getByTestId('errorIcon')).toBeInTheDocument();
      expect(screen.getByText('errorMessage')).toBeInTheDocument();
    });

    and('LabelInput should have the error style', () => {
      expect(container).toMatchSnapshot();
    });
  });
});
