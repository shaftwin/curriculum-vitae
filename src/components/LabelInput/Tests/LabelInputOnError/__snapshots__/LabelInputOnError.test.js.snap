// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelInputOnError Default LabelInput on error without message 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-size: 0.813rem;
  width: 16.625rem;
}

.emotion-1 {
  margin-bottom: 0.25rem;
  font-weight: 600;
  color: #F55858;
}

.emotion-2 {
  border-radius: 0.25rem;
  border: 0.063rem solid #F55858;
  color: #676478;
  cursor: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0rem 0.75rem;
}

.emotion-2:hover:not([disabled]) {
  border: #EBEEF0;
  box-shadow: 0rem 0rem 0.313rem #F1F1F1;
  color: #A4AAC0;
}

.emotion-2:focus-within {
  border: 0.063rem solid #00DBC7;
  box-shadow: 0rem 0rem 0.875rem #F1F1F1;
  color: #676478;
}

.emotion-3 {
  outline: none;
  border: 0;
  background-color: inherit;
  color: inherit;
  width: 100%;
  font-size: 0.813rem;
  min-height: 2rem;
  cursor: inherit;
}

.emotion-4 {
  position: relative;
  left: unset;
}

<div>
  <div
    class="emotion-0"
    data-testid="LabelInput"
  >
    <label
      class="emotion-1"
      data-testid="label"
    >
      testLabel
    </label>
    <div
      class="emotion-2"
      data-testid="inputWrapper"
    >
      <input
        class="emotion-3"
        data-testid="input"
        type="text"
        value="testValue"
      />
      <div
        class="emotion-4"
        data-testid="errorIcon"
      >
        <svg
          fill="none"
          height="20"
          viewBox="0 0 21 20"
          width="21"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="10.207"
            cy="10"
            fill="#F55858"
            opacity="0.2"
            r="10"
          />
          <path
            d="M9.89165 11.48L9.39965 7.98802V5.62402H11.0317V7.98802L10.5637 11.48H9.89165ZM10.2157 14.144C9.87165 14.144 9.62365 14.064 9.47165 13.904C9.31965 13.744 9.24365 13.54 9.24365 13.292V13.052C9.24365 12.804 9.31965 12.6 9.47165 12.44C9.62365 12.28 9.87165 12.2 10.2157 12.2C10.5517 12.2 10.7957 12.28 10.9477 12.44C11.0997 12.6 11.1757 12.804 11.1757 13.052V13.292C11.1757 13.54 11.0997 13.744 10.9477 13.904C10.7957 14.064 10.5517 14.144 10.2157 14.144Z"
            fill="#F55858"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`LabelInputOnError LabelInput throw error with message 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-size: 0.813rem;
  width: 16.625rem;
}

.emotion-1 {
  margin-bottom: 0.25rem;
  font-weight: 600;
  color: #F55858;
}

.emotion-2 {
  border-radius: 0.25rem;
  border: 0.063rem solid #F55858;
  color: #676478;
  cursor: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0rem 0.75rem;
}

.emotion-2:hover:not([disabled]) {
  border: #EBEEF0;
  box-shadow: 0rem 0rem 0.313rem #F1F1F1;
  color: #A4AAC0;
}

.emotion-2:focus-within {
  border: 0.063rem solid #00DBC7;
  box-shadow: 0rem 0rem 0.875rem #F1F1F1;
  color: #676478;
}

.emotion-3 {
  outline: none;
  border: 0;
  background-color: inherit;
  color: inherit;
  width: 100%;
  font-size: 0.813rem;
  min-height: 2rem;
  cursor: inherit;
}

.emotion-4 {
  position: relative;
  left: unset;
}

.emotion-5 {
  margin-bottom: 0.25rem;
  font-weight: 600;
  color: #F55858;
  margin: 0.25rem 0rem 0rem 0rem;
  font-size: 0.625rem;
}

<div>
  <div
    class="emotion-0"
    data-testid="LabelInput"
  >
    <label
      class="emotion-1"
      data-testid="label"
    >
      testLabel
    </label>
    <div
      class="emotion-2"
      data-testid="inputWrapper"
    >
      <input
        class="emotion-3"
        data-testid="input"
        type="text"
        value="testValue"
      />
      <div
        class="emotion-4"
        data-testid="errorIcon"
      >
        <svg
          fill="none"
          height="20"
          viewBox="0 0 21 20"
          width="21"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="10.207"
            cy="10"
            fill="#F55858"
            opacity="0.2"
            r="10"
          />
          <path
            d="M9.89165 11.48L9.39965 7.98802V5.62402H11.0317V7.98802L10.5637 11.48H9.89165ZM10.2157 14.144C9.87165 14.144 9.62365 14.064 9.47165 13.904C9.31965 13.744 9.24365 13.54 9.24365 13.292V13.052C9.24365 12.804 9.31965 12.6 9.47165 12.44C9.62365 12.28 9.87165 12.2 10.2157 12.2C10.5517 12.2 10.7957 12.28 10.9477 12.44C11.0997 12.6 11.1757 12.804 11.1757 13.052V13.292C11.1757 13.54 11.0997 13.744 10.9477 13.904C10.7957 14.064 10.5517 14.144 10.2157 14.144Z"
            fill="#F55858"
          />
        </svg>
      </div>
    </div>
    <label
      class="emotion-5"
      data-testid="errorMessage"
    >
      errorMessage
    </label>
  </div>
</div>
`;
